File: scraper/email_renderer.py
"""Renders the weekly job digest as clean HTML and plain-text."""
import datetime


def _source_badge(source_name: str) -> str:
    return (
        f'<span style="display:inline-block;font-family:monospace;font-size:11px;'
        f'background:#27272a;color:#a1a1aa;padding:1px 6px;border-radius:4px;'
        f'white-space:nowrap">{source_name}</span>'
    )


def _job_row(listing: dict) -> str:
    role = listing.get('role_title', '')
    company = listing.get('company', '')
    url = listing.get('listing_url', '#')
    source = listing.get('source_name', '')
    tags = listing.get('tags') or []
    tag_html = ''
    if tags:
        pills = ''.join(
            f'<span style="display:inline-block;font-size:11px;background:#3f3f46;'
            f'color:#d4d4d8;padding:1px 7px;border-radius:999px;margin-right:4px">'
            f'{t}</span>'
            for t in tags[:4]
        )
        tag_html = f'<div style="margin-top:3px">{pills}</div>'

    return f'''
<tr>
  <td style="padding:10px 0;border-bottom:1px solid #27272a;vertical-align:top">
    <a href="{url}" style="font-size:14px;font-weight:600;color:#a78bfa;text-decoration:none"
       target="_blank">{role}</a>
    <span style="color:#71717a;font-size:13px;margin-left:8px">{company}</span>
    <div style="margin-top:4px">{_source_badge(source)}{tag_html}</div>
  </td>
</tr>'''


def _section(title: str, listings: list) -> str:
    if not listings:
        return ''
    rows = ''.join(_job_row(l) for l in listings)
    return f'''
<tr>
  <td style="padding:20px 0 6px">
    <div style="font-size:11px;font-weight:700;letter-spacing:.08em;text-transform:uppercase;
                color:#71717a;border-bottom:1px solid #3f3f46;padding-bottom:6px">{title}</div>
  </td>
</tr>
{rows}'''


def render_html(
    new_listings: list,
    week_key: str,
    total_scraped: int,
    new_count: int,
    dupe_count: int,
    filtered_count: int,
) -> str:
    """Return a complete HTML email for the weekly digest."""
    curated = [l for l in new_listings if l.get('is_filtered_in')]
    other = [l for l in new_listings if not l.get('is_filtered_in')]

    # Split curated by source type
    vc_curated = [l for l in curated if l.get('source_name') not in _NEWSLETTER_NAMES]
    nl_curated = [l for l in curated if l.get('source_name') in _NEWSLETTER_NAMES]

    curated_section = ''
    if vc_curated:
        curated_section += _section('VC Board Picks', vc_curated)
    if nl_curated:
        curated_section += _section('Newsletter Picks', nl_curated)

    other_preview = other[:30]  # cap non-curated at 30
    other_section = _section('All Other New Listings', other_preview) if other_preview else ''
    more_note = ''
    if len(other) > 30:
        more_note = (
            f'<tr><td style="padding:8px 0;color:#71717a;font-size:12px">'
            f'…and {len(other) - 30} more. '
            f'<a href="https://gdqfekecdtkiwwmyjwfz.supabase.co" style="color:#a78bfa">View all →</a>'
            f'</td></tr>'
        )

    run_date = datetime.datetime.now().strftime('%B %d, %Y')
    week_label = _week_label(week_key)

    html = f'''<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>Startup Jobs Digest — {week_label}</title>
</head>
<body style="margin:0;padding:0;background:#09090b;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif">
<table width="100%" cellpadding="0" cellspacing="0" style="background:#09090b">
<tr><td align="center" style="padding:32px 16px">

  <table width="600" cellpadding="0" cellspacing="0"
         style="background:#18181b;border-radius:10px;border:1px solid #27272a;max-width:600px;width:100%">

    <!-- Header -->
    <tr>
      <td style="padding:28px 32px 20px;border-bottom:1px solid #27272a">
        <div style="font-size:20px;font-weight:700;color:#fafafa">Startup Jobs Digest</div>
        <div style="font-size:13px;color:#71717a;margin-top:4px">{week_label} &middot; {run_date}</div>
      </td>
    </tr>

    <!-- Stats bar -->
    <tr>
      <td style="padding:14px 32px;background:#09090b;border-bottom:1px solid #27272a">
        <table width="100%" cellpadding="0" cellspacing="0">
          <tr>
            {_stat('Scraped', total_scraped)}
            {_stat('New', new_count, '#a78bfa')}
            {_stat('Curated', filtered_count, '#fbbf24')}
            {_stat('Dupes skipped', dupe_count)}
          </tr>
        </table>
      </td>
    </tr>

    <!-- Body -->
    <tr>
      <td style="padding:8px 32px 24px">
        <table width="100%" cellpadding="0" cellspacing="0">
          {curated_section}
          {other_section}
          {more_note}
        </table>
      </td>
    </tr>

    <!-- Footer -->
    <tr>
      <td style="padding:18px 32px;border-top:1px solid #27272a;font-size:11px;color:#52525b;text-align:center">
        Sent by the Startup Job Aggregator pipeline &middot;
        <a href="mailto:databelmont@gmail.com" style="color:#52525b">databelmont@gmail.com</a>
      </td>
    </tr>

  </table>
</td></tr>
</table>
</body>
</html>'''
    return html


def render_plain(
    new_listings: list,
    week_key: str,
    total_scraped: int,
    new_count: int,
    dupe_count: int,
    filtered_count: int,
) -> str:
    """Return a plain-text fallback for the digest."""
    week_label = _week_label(week_key)
    lines = [
        f'STARTUP JOBS DIGEST — {week_label}',
        '=' * 50,
        f'Scraped: {total_scraped}  |  New: {new_count}  |  Curated: {filtered_count}  |  Dupes: {dupe_count}',
        '',
    ]
    curated = [l for l in new_listings if l.get('is_filtered_in')]
    if curated:
        lines.append('--- CURATED PICKS ---')
        for l in curated:
            lines.append(f"[{l.get('source_name','')}] {l.get('role_title','')} @ {l.get('company','')}")
            lines.append(f"  {l.get('listing_url','')}")
        lines.append('')
    other = [l for l in new_listings if not l.get('is_filtered_in')][:30]
    if other:
        lines.append('--- OTHER NEW LISTINGS ---')
        for l in other:
            lines.append(f"[{l.get('source_name','')}] {l.get('role_title','')} @ {l.get('company','')}")
            lines.append(f"  {l.get('listing_url','')}")
    return '\n'.join(lines)


# --- helpers ---

_NEWSLETTER_NAMES = {
    'Job Hunting Sux', 'Ali Rohde Jobs', 'Next Play',
    'Nonlinear Techies', 'Beyond Bay St', 'Why You Should Join', 'Web of Jobs',
}


def _stat(label: str, value: int, color: str = '#a1a1aa') -> str:
    return (
        f'<td align="center" style="padding:0 12px">'
        f'<div style="font-size:20px;font-weight:700;color:{color}">{value}</div>'
        f'<div style="font-size:10px;color:#52525b;text-transform:uppercase;letter-spacing:.06em">{label}</div>'
        f'</td>'
    )


def _week_label(week_key: str) -> str:
    """'2026-W11' → 'Week of Mar 9, 2026'"""
    try:
        year, w = week_key.split('-W')
        year, w = int(year), int(w)
        jan4 = datetime.date(year, 1, 4)
        monday_w1 = jan4 - datetime.timedelta(days=jan4.weekday())
        monday = monday_w1 + datetime.timedelta(weeks=w - 1)
        return 'Week of ' + monday.strftime('%b %-d, %Y')
    except Exception:
        return week_key

File: src/lib/emailRenderer.ts
import type { Job } from '@/types';

const NEWSLETTER_NAMES = new Set([
  'Job Hunting Sux', 'Ali Rohde Jobs', 'Next Play',
  'Nonlinear Techies', 'Beyond Bay St', 'Why You Should Join', 'Web of Jobs',
]);

function weekLabel(weekKey: string): string {
  try {
    const [yearStr, wStr] = weekKey.split('-W');
    const year = parseInt(yearStr, 10);
    const week = parseInt(wStr, 10);
    const jan4 = new Date(year, 0, 4);
    const monday1 = new Date(jan4);
    monday1.setDate(jan4.getDate() - ((jan4.getDay() + 6) % 7));
    const monday = new Date(monday1);
    monday.setDate(monday1.getDate() + (week - 1) * 7);
    return 'Week of ' + monday.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
  } catch {
    return weekKey;
  }
}

function formatDate(d: string | null): string {
  if (!d) return 'Unknown';
  const dt = new Date(d);
  if (isNaN(dt.getTime())) return 'Unknown';
  return dt.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
}

// Full card for curated jobs — highlighted with violet left border + elevated background
function curatedCard(job: Job): string {
  const tags = (job.tags ?? []).slice(0, 5)
    .map(t => `<span style="display:inline-block;font-size:11px;background:#2e1065;color:#c4b5fd;padding:2px 8px;border-radius:999px;margin-right:4px;margin-top:3px">${t}</span>`)
    .join('');
  const snippet = job.raw_snippet
    ? `<div style="font-size:12px;color:#71717a;font-style:italic;margin-top:8px;border-left:2px solid #3f3f46;padding-left:8px;line-height:1.5">${job.raw_snippet.slice(0, 200)}${job.raw_snippet.length > 200 ? '…' : ''}</div>`
    : '';

  return `
<tr>
  <td style="padding:6px 0">
    <table width="100%" cellpadding="0" cellspacing="0" style="background:#1a1625;border:1px solid #3b2f6b;border-left:3px solid #7c3aed;border-radius:6px">
      <tr><td style="padding:14px 16px">
        <table width="100%" cellpadding="0" cellspacing="0"><tr>
          <td style="vertical-align:top">
            <a href="${job.listing_url}" style="font-size:15px;font-weight:700;color:#c4b5fd;text-decoration:none;line-height:1.3" target="_blank">${job.role_title}</a>
            <div style="font-size:13px;color:#a1a1aa;margin-top:3px;font-weight:500">${job.company}</div>
          </td>
          <td style="vertical-align:top;text-align:right;white-space:nowrap;padding-left:12px">
            <span style="display:inline-block;font-size:10px;font-weight:700;letter-spacing:.06em;text-transform:uppercase;background:#2e1065;color:#a78bfa;padding:3px 8px;border-radius:4px">Curated</span>
          </td>
        </tr></table>
        <div style="margin-top:8px">
          <span style="display:inline-block;font-family:monospace;font-size:11px;background:#27272a;color:#a1a1aa;padding:2px 7px;border-radius:4px">${job.source_name}</span>
          <span style="font-size:11px;color:#52525b;margin-left:8px">${formatDate(job.date_scraped)}</span>
        </div>
        ${tags ? `<div style="margin-top:4px">${tags}</div>` : ''}
        ${snippet}
      </td></tr>
    </table>
  </td>
</tr>`;
}

// Compact single-line row for non-curated jobs
function compactRow(job: Job): string {
  return `
<tr>
  <td style="padding:5px 0;border-bottom:1px solid #1f1f23">
    <table width="100%" cellpadding="0" cellspacing="0"><tr>
      <td style="vertical-align:middle">
        <a href="${job.listing_url}" style="font-size:13px;color:#a1a1aa;text-decoration:none;font-weight:500" target="_blank">${job.role_title}</a>
        <span style="color:#52525b;font-size:12px;margin-left:6px">${job.company}</span>
      </td>
      <td style="vertical-align:middle;text-align:right;white-space:nowrap;padding-left:8px">
        <span style="font-family:monospace;font-size:10px;color:#3f3f46">${job.source_name}</span>
      </td>
    </tr></table>
  </td>
</tr>`;
}

function sectionHeader(title: string, count: number, accent = '#71717a'): string {
  return `
<tr>
  <td style="padding:24px 0 10px">
    <table width="100%" cellpadding="0" cellspacing="0"><tr>
      <td style="border-bottom:1px solid #3f3f46;padding-bottom:8px">
        <span style="font-size:11px;font-weight:700;letter-spacing:.08em;text-transform:uppercase;color:${accent}">${title}</span>
        <span style="font-size:11px;color:#3f3f46;margin-left:8px">${count}</span>
      </td>
    </tr></table>
  </td>
</tr>`;
}

function statCell(label: string, value: number, color = '#a1a1aa'): string {
  return `
<td align="center" style="padding:0 16px">
  <div style="font-size:22px;font-weight:700;color:${color}">${value}</div>
  <div style="font-size:10px;color:#52525b;text-transform:uppercase;letter-spacing:.06em">${label}</div>
</td>`;
}

export function renderDigestHtml(jobs: Job[], weekKey: string): string {
  const curated = jobs.filter(j => j.is_filtered_in);
  const other = jobs.filter(j => !j.is_filtered_in);
  const vcCurated = curated.filter(j => !NEWSLETTER_NAMES.has(j.source_name));
  const nlCurated = curated.filter(j => NEWSLETTER_NAMES.has(j.source_name));
  const otherPreview = other.slice(0, 40);

  const runDate = new Date().toLocaleDateString('en-US', { month: 'long', day: 'numeric', year: 'numeric' });

  const vcSection = vcCurated.length === 0 ? '' : `
    ${sectionHeader('VC Board Picks', vcCurated.length, '#a78bfa')}
    ${vcCurated.map(curatedCard).join('')}`;

  const nlSection = nlCurated.length === 0 ? '' : `
    ${sectionHeader('Newsletter Picks', nlCurated.length, '#a78bfa')}
    ${nlCurated.map(curatedCard).join('')}`;

  const otherSection = otherPreview.length === 0 ? '' : `
    ${sectionHeader('All Other Listings', other.length)}
    ${otherPreview.map(compactRow).join('')}
    ${other.length > 40 ? `<tr><td style="padding:8px 0 0;font-size:12px;color:#3f3f46">…and ${other.length - 40} more</td></tr>` : ''}`;

  return `<!DOCTYPE html>
<html lang="en">
<head><meta charset="UTF-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>Startup Jobs Digest — ${weekLabel(weekKey)}</title></head>
<body style="margin:0;padding:0;background:#09090b;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif">
<table width="100%" cellpadding="0" cellspacing="0" style="background:#09090b">
<tr><td align="center" style="padding:32px 16px">
<table width="600" cellpadding="0" cellspacing="0" style="background:#18181b;border-radius:10px;border:1px solid #27272a;max-width:600px;width:100%">

  <!-- Header -->
  <tr><td style="padding:28px 32px 20px;border-bottom:1px solid #27272a">
    <div style="font-size:20px;font-weight:700;color:#fafafa">Startup Jobs Digest</div>
    <div style="font-size:13px;color:#71717a;margin-top:4px">${weekLabel(weekKey)} &middot; ${runDate}</div>
  </td></tr>

  <!-- Stats bar -->
  <tr><td style="padding:14px 32px;background:#09090b;border-bottom:1px solid #27272a">
    <table width="100%" cellpadding="0" cellspacing="0"><tr>
      ${statCell('Total', jobs.length)}
      ${statCell('Curated', curated.length, '#a78bfa')}
      ${statCell('VC Boards', vcCurated.length, '#fbbf24')}
      ${statCell('Newsletters', nlCurated.length, '#34d399')}
    </tr></table>
  </td></tr>

  <!-- Body -->
  <tr><td style="padding:4px 32px 28px">
    <table width="100%" cellpadding="0" cellspacing="0">
      ${vcSection}
      ${nlSection}
      ${otherSection}
    </table>
  </td></tr>

  <!-- Footer -->
  <tr><td style="padding:18px 32px;border-top:1px solid #27272a;font-size:11px;color:#52525b;text-align:center">
    Sent by Startup Job Aggregator &middot; <a href="mailto:databelmont@gmail.com" style="color:#52525b">databelmont@gmail.com</a>
  </td></tr>

</table>
</td></tr>
</table>
</body>
</html>`;
}

export function renderDigestPlain(jobs: Job[], weekKey: string): string {
  const curated = jobs.filter(j => j.is_filtered_in);
  const other = jobs.filter(j => !j.is_filtered_in);
  const lines = [
    `STARTUP JOBS DIGEST — ${weekLabel(weekKey)}`,
    '='.repeat(50),
    `Total: ${jobs.length}  |  Curated: ${curated.length}  |  Other: ${other.length}`,
    '',
  ];
  if (curated.length > 0) {
    lines.push('★ CURATED PICKS');
    lines.push('-'.repeat(40));
    for (const j of curated) {
      lines.push(`${j.role_title} @ ${j.company}`);
      lines.push(`  Source: ${j.source_name}  |  Tags: ${(j.tags ?? []).join(', ') || 'none'}`);
      lines.push(`  ${j.listing_url}`);
      lines.push('');
    }
  }
  if (other.length > 0) {
    lines.push('ALL OTHER LISTINGS');
    lines.push('-'.repeat(40));
    for (const j of other.slice(0, 40)) {
      lines.push(`${j.role_title} @ ${j.company}  [${j.source_name}]`);
      lines.push(`  ${j.listing_url}`);
    }
    if (other.length > 40) lines.push(`…and ${other.length - 40} more`);
  }
  return lines.join('\n');
}
